### ISSUE TITLE
messageEvent.ports[0] returns an invalid type

### ISSUE THREAD

**Issue by terjanq** [microsoft/TypeScript-DOM-lib-generator#1385]
When handling message events the type for `messageEvent.ports[0]` is `MessagePort` instead of the desired `MessagePort | undefined`. It can be undefined because the [transferable](https://developer.mozilla.org/en-US/docs/Glossary/Transferable_objects) might not contain any ports.

[PoC](https://www.typescriptlang.org/play?#code/O4SwdgJg9sB0CGEIFEBuBTMAXAMiAzlpugE4AUA5ALbr77wDm6FANAARnoBcbAsrfSZpMWAJRsAvAD42AbwBQbNgGMoYfFAA26WJqgNKEAgAdN8AJ74KogNyK2O41BJZ8AbQAMAXVhPC-OkZ0MgBGWzYAegiVKCpjEG18dgAjAFcsNioQBgALDOSSdHgAazYsHPQYiEr4DJJUsABaLBAae1V1LR09AwpoWjYwKAyjfFMLazsAX1t5eT8sAMFgkIAmAGZbIA)

```typescript
window.addEventListener('message', (messageEvent: MessageEvent) => {
  console.log('displays');
  messageEvent.ports[0].postMessage(1); // compiles, but might break the code at run-time
  console.log('does not display');
});
postMessage(123);
```

**Comment by terjanq** [comment]
Yeah, I guess it's a problem with the arrays. It's harder to spot issues where the developer needs to know all of that stuff.

**Comment by HolgerJeromin** [comment]
You can add `"noUncheckedIndexedAccess": true` to your tsconfig to catch such errors.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#checked-indexed-accesses---nouncheckedindexedaccess
https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess

https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAhgLhgQQE6rgTwDxgK4C2ARgKaoB8MAvDANoC6AUKJLEdfLQIxMx8D0-GAD0A-EA

**Comment by terjanq** [comment]
Thanks. I guess it can be closed now as indeed it's a wider issue with Arrays.
